google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java

```diff
@@ -827,6 +855,22 @@ public Builder setEndpoint(String endpoint) {
       return this;
     }
 
+    /** Gives the ability to enable transport compression. */
+    public Builder setEnableCompression(boolean enableCompression) {
+      this.enableCompression = enableCompression;
+      return this;
+    }
+
+    /**
+     * Sets the threshold (in bytes) above which messages are compressed for transport. Only takes
+     * effect if setEnableCompression(true) is also called."
+     */
+    public Builder setCompressionBytesThreshold(long compressionBytesThreshold) {
+      this.enableCompressionBytesThreshold = true;
```

Comment:
Another way we could do this is to have a precondition check here. Something like:
```
public Builder setCompressionBytesThreshold(long compressionBytesThreshold) {
      Preconditions.checkArgument(this.enableCompressionBytesThreshold);
      this.compressionBytesThreshold = compressionBytesThreshold;
      return this;
    }
```

Which would then require the builder to be used in a certain order:
```
// This would work
publisher = Publisher.newBuilder(topicName)
                            .setEnableCompression(true)
                            .setCompressionBytesThreshold(500)
                            .build();
                            
// This would throw an exception
publisher = Publisher.newBuilder(topicName)
                            .setCompressionBytesThreshold(500)
                            .setEnableCompression(true)
                            .build();

// This would also throw an exception
publisher = Publisher.newBuilder(topicName)
                            .setCompressionBytesThreshold(500)
                            .build();
```

Reply:
I honestly don't think enforcing the order of the call in the the builder is a good thing. @kamalaboulhosn Please share your thoughts.

Reply:
The issue is that one needs to recreate the Builder entirely or basically rewrite their code. Maybe this isn't much of an issue, but I'm worried about code that has something like:

```
Publisher.Builder builder = Publisher.newBuilder(topicName);

builder.setEnableCompression(true).setCompressionBytesThredshold(500);

if (dontCompressInTest) {
  builder.setEnableCompression(false);
}
```

And what if the `dontCompressInTest` check is in another method or something in another class (I have used this model myself in places where I override a class that creates  Publisher and passes the bulider around to set additional properties).

Yes, they would get an error when they call build. But why do they need to get an error here? This seems like a unique behavior in the builder where the order in which things are set produces different results. That's what I want to avoid. One option is to reset `enableCompressionBytesThreshold` to `false` if one calls `setEnabledCompression(false)`. Alternatively, as I suggested, just let the documentation cover this case and don't have the check in `build`.